astropy/io/fits/hdu/base.py

```diff
@@ -1464,55 +1464,23 @@ def _compute_checksum(self, data, sum32=0):
         -------
         ones complement checksum
         """
-        blocklen = 2880
-        sum32 = np.uint32(sum32)
-        for i in range(0, len(data), blocklen):
-            length = min(blocklen, len(data) - i)  # ????
-            sum32 = self._compute_hdu_checksum(data[i : i + length], sum32)
-        return sum32
-
-    def _compute_hdu_checksum(self, data, sum32=0):
-        """
-        Translated from FITS Checksum Proposal by Seaman, Pence, and Rots.
-        Use uint32 literals as a hedge against type promotion to int64.
-
-        This code should only be called with blocks of 2880 bytes
-        Longer blocks result in non-standard checksums with carry overflow
-        Historically,  this code *was* called with larger blocks and for that
-        reason still needs to be for backward compatibility.
-        """
-        u8 = np.uint32(8)
-        u16 = np.uint32(16)
-        uFFFF = np.uint32(0xFFFF)
-
-        if data.nbytes % 2:
-            last = data[-1]
-            data = data[:-1]
-        else:
-            last = np.uint32(0)
-
-        data = data.view(">u2")
-
-        hi = sum32 >> u16
-        lo = sum32 & uFFFF
-        hi += np.add.reduce(data[0::2], dtype=np.uint64)
-        lo += np.add.reduce(data[1::2], dtype=np.uint64)
-
-        if (data.nbytes // 2) % 2:
-            lo += last << u8
-        else:
-            hi += last << u8
-
-        hicarry = hi >> u16
-        locarry = lo >> u16
-
-        while hicarry or locarry:
-            hi = (hi & uFFFF) + locarry
-            lo = (lo & uFFFF) + hicarry
-            hicarry = hi >> u16
-            locarry = lo >> u16
-
-        return (hi << u16) + lo
+        # Ensures no overflow in uint64 sum (logically, this should perhaps be
+        # 1 << 34, since we care about the number of uint32 data, not the
+        # number of bytes, but, heck, better safe than sorry.
+        blocklen = 1 << 32
+        # The cast to uint32 is not needed by the tests, and seems odd, since
+        # we explicitly use higher bits otherwise. Kept regardless since it
+        # was there before.
+        s = int(np.uint32(sum32))
+        for piece in np.split(data, range(blocklen, len(data), blocklen)):
+            if extra := piece.nbytes % 4:
+                last = bytes(piece[-extra:]) + b"\00" * (4 - extra)
+                s += int.from_bytes(last, byteorder="big", signed=False)
+                piece = piece[:-extra]
+            s += int(piece.view(">u4").sum(dtype="u8"))
```

Comment:
Just curious, would there be any performance benefit to doing the view and sum in a vectorised way or would it be tricky because it would require padding the array first to be a multiple of the chunk size?